pinot-plugins/pinot-stream-ingestion/pinot-kinesis/src/main/java/org/apache/pinot/plugin/stream/kinesis/KinesisStreamMetadataProvider.java

```diff
@@ -189,6 +192,24 @@ private boolean consumedEndOfShard(StreamPartitionMsgOffset startCheckpoint,
   }
 
   @Override
-  public void close() {
+  public Map<String, PartitionLagState> getCurrentPartitionLagState(
+      Map<String, ConsumerPartitionState> currentPartitionStateMap) {
+    Map<String, PartitionLagState> perPartitionLag = new HashMap<>();
+    for (Map.Entry<String, ConsumerPartitionState> entry: currentPartitionStateMap.entrySet()) {
+      ConsumerPartitionState partitionState = entry.getValue();
+      // Compute record-availability
+      String recordAvailabilityLag = "UNKNOWN";
+      RowMetadata lastProcessedMessageMetadata = partitionState.getLastProcessedRowMetadata();
+      if (lastProcessedMessageMetadata != null && partitionState.getLastProcessedTimeMs() > 0) {
+        long availabilityLag = partitionState.getLastProcessedTimeMs()
+            - lastProcessedMessageMetadata.getRecordIngestionTimeMs();
+        recordAvailabilityLag = String.valueOf(availabilityLag);
+      }
+      perPartitionLag.put(entry.getKey(), new KinesisConsumerPartitionLag(recordAvailabilityLag));
```

Comment:
IMO, we should be using this metric. Kinesis official dashboards also use the same metric
[image](https://user-images.githubusercontent.com/5559703/235164581-f9c3a557-b425-4bc6-a4c0-759e76e082e0.png)

Reply:
Also, we should add a javadoc here declaring what happens to lag in case of a reshard operation.

Reply:
I agree that the kinesis official metric will be the most accurate one. Let me re-think if this approach makes sense because we now have a mechanism to compute ingestionDelay in the realtime engine itself based on this apprx arrival time. That is ideally what we should be using uniformly across all connectors. 

> Also, we should add a javadoc here declaring what happens to lag in case of a reshard operation.

Good point. When re-shard happens, I am guessing one or more of the partitions may no longer have any data being ingested and cause this lag to continuously increase, leading to false alarms. Is my understanding right?

Reply:
Yes, the understanding is correct here. We do remove the parent partitions when we have consumed all data from them. Maybe that fixes the issues?

Reply:
yea it won't be in the currentPartitionStateMap anymore once we've moved on to the children, so APIs should be fine. Do we also emit any metrics per partition about this? If yes, those will face issues of false alerts

Reply:
Yeah. So, the metrics that is actually used for monitoring is `ingestionDelayMs`. This PR provides an API to compute the point in time ingestion lag (This API predates the ingestion lag) and doesn't actually create any metric. So, I think we don't need to worry about false alerts here.
We do enable the `ingestionDelayLagMs` metric at a partition level. I am not entirely sure how to remove a metric emitted by the realtime engine, when the connector manages the set of active partitions.. Need to look at the code again. but this problem should is out of scope for this PR.